shared/src/main/scala/mlscript/TypeDefs.scala

```diff
@@ -555,89 +555,92 @@ class TypeDefs extends ConstraintSolver { self: Typer =>
-      case TraitTag(_) | ClassTag(_, _) => ()
-      case ExtrType(pol) => ()
-      case t: TypeVariable => tyDef.tvarVariance.updateWith(t) {
-        case None => Some(curVariance)
-        case Some(value) =>
-          val newVariance = value && curVariance
-          if (newVariance =/= curVariance) varianceUpdated = true;
-          Some(newVariance)
-      })
-      case RecordType(fields) => fields.foreach {
-        case (_ , fieldTy) => fieldVarianceHelper(fieldTy, curVariance)
-      }
-      case NegType(negated) =>
-        updateVariance(negated, curVariance.flip(VarianceInfo.contra()))
-      case TypeRef(defn, targs) =>
-        val typeRefDef = ctx.tyDefs.getOrElse(defn.name, throw new Exception(s"Cannot find $defn type definition in context"))
-        targs.zip(typeRefDef.tvars).foreach { case (targ, tvar) =>
-          val tvarVariance = if (curVariance === VarianceInfo.contra())
-              typeRefDef.tvarVariance.getOrElse(tvar, VarianceInfo.bi()).flip(VarianceInfo.contra())
-            else
-              typeRefDef.tvarVariance.getOrElse(tvar, VarianceInfo.bi()).flip(VarianceInfo.co())
-          updateVariance(targ, tvarVariance)
-        }
-      case ComposedType(pol, lhs, rhs) =>
-        updateVariance(lhs, curVariance)
-        updateVariance(rhs, curVariance)
-      case TypeBounds(lb, ub) =>
-        updateVariance(lb, curVariance.flip(VarianceInfo.contra()))
-        updateVariance(ub, curVariance.flip(VarianceInfo.co()))
-      case ArrayType(inner) => fieldVarianceHelper(inner, curVariance)
-      case TupleType(fields) => fields.foreach {
-          case (_ , fieldTy) => fieldVarianceHelper(fieldTy, curVariance)
+    def updateVariance(ty: SimpleType, curVariance: VarianceInfo)(implicit ctx: Ctx, tyDef: TypeDef): Unit =
+      trace(s"updateVariance(ty: $ty of ${ty.getClass()}, curVariance: $curVariance)(${tyDef.nme.name}.${tyDef.tvarVariances})") {
+        ty match {
+          case ProxyType(underlying) => updateVariance(underlying, curVariance)
+          case TraitTag(_) | ClassTag(_, _) => ()
+          case ExtrType(pol) => ()
+          case t: TypeVariable => val _ = tyDef.tvarVariances.updateWith(t) {
+            case None => Some(curVariance)
+            case Some(value) =>
+              val newVariance = value && curVariance
+              if (newVariance =/= curVariance) varianceUpdated = true;
+              Some(newVariance)
+          }
+          case RecordType(fields) => fields.foreach {
+            case (_ , fieldTy) => fieldVarianceHelper(fieldTy, curVariance)
+          }
+          case NegType(negated) =>
+            updateVariance(negated, curVariance.flip)
+          case TypeRef(defn, targs) =>
+            val typeRefDef = ctx.tyDefs.getOrElse(defn.name, throw new Exception(s"Cannot find $defn type definition in context"))
+            targs.zip(typeRefDef.tvars).foreach { case (targ, tvar) =>
+              typeRefDef.tvarVariances.getOrElse(tvar, VarianceInfo.bi) match {
```

Comment:
```suggestion
              typeRefDef.tvarVariances.getOrElseUpdate(tvar, VarianceInfo.bi) match {
```

Please be more careful when reviewing my comments and suggestions. This is the second one you partially disregarded here, which I take as a lack of respect for my time spent reviewing your changes.

(You can ignore this if you plan to but simply haven't gotten to making the change yet – but then the comment should not have been marked resolved, because you may forget to do it and I will not necessarily find out!)

Reply:
What's the reasoning for reverting this back yet again?

Reply:
Actually I realize I should do the computation inside the option monad using `foreach`. I'm pre-setting all type variables of a type definition to bi-variance, so the invariant logic here is that the variance for the type variable for the will always exist, otherwise just ignore it.

Reply:
Feels like the GDP technique could be useful here :grin:. We know that values for these keys are set, but still have to handle the else case or use for each.

Reply:
I'll resolve this now, using the `foreach`. Using `foreach` in the above line also solves the error that was showing up in `BadAlias`. I'll consider this resolved then.